From: zhengbin <zhengbin13@huawei.com>
Date: Tue, 23 Jul 2019 22:10:42 +0800
Subject: blk-mq: Fix memory leak in blk_mq_init_allocated_queue error handling
Patch-mainline: v5.4-rc1
Git-commit: 73d9c8d4c0017e21e1ff519474ceb1450484dc9a
References: bsc#1151610

If blk_mq_init_allocated_queue->elevator_init_mq fails, need to release
the previously requested resources.

Fixes: d34849913819 ("blk-mq-sched: allow setting of default IO scheduler")
Signed-off-by: zhengbin <zhengbin13@huawei.com>
Signed-off-by: Jens Axboe <axboe@kernel.dk>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 block/blk-mq.c |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2403,6 +2403,8 @@ static void blk_mq_realloc_hw_ctxs(struc
 struct request_queue *blk_mq_init_allocated_queue(struct blk_mq_tag_set *set,
 						  struct request_queue *q)
 {
+	int ret = -ENOMEM;
+
 	/* mark the queue as mq asap */
 	q->mq_ops = set->ops;
 
@@ -2468,17 +2470,18 @@ struct request_queue *blk_mq_init_alloca
 	blk_mq_map_swqueue(q);
 
 	if (!(set->flags & BLK_MQ_F_NO_SCHED)) {
-		int ret;
-
 		ret = elevator_init_mq(q);
 		if (ret)
-			return ERR_PTR(ret);
+			goto err_tag_set;
 	}
 
 	return q;
 
+err_tag_set:
+	blk_mq_del_queue_tag_set(q);
 err_hctxs:
 	kfree(q->queue_hw_ctx);
+	q->nr_hw_queues = 0;
 err_percpu:
 	free_percpu(q->queue_ctx);
 err_exit:

